tests/testRobot.cpp

```diff
@@ -231,6 +237,31 @@ TEST(ForwardKinematics, FourBar) {
   THROWS_EXCEPTION(robot.forwardKinematics(wrong_vels));
 }
 
+TEST(Robot, Equality) {
+  Robot robot1 = CreateRobotFromFile(
+      kSdfPath + std::string("test/four_bar_linkage_pure.sdf"));
+  Robot robot2 = CreateRobotFromFile(
+      kSdfPath + std::string("test/four_bar_linkage_pure.sdf"));
+
+  EXPECT(robot1 == robot2);
+  EXPECT(robot1.equals(robot2));
+}
```

Comment:
sorry I don't want to drag this out longer than it needs to, but could you add a false-y test?  e.g. `robot1.joint1.parent = nullptr; EXPECT(robot1 != robot2)`

I'm personally just a little suspicious of the circular reference serialization thing so want to make sure the round-trip serialization unit tests will cover incorrect serializations (e.g. roundtrip test will always return true if `bool Robot::equals(Robot other) { return true; }`)

Reply:
Good point. Please drag this on as long as needed so this PR is amazing. :slightly_smiling_face:

Reply:
Hmm there's no clean way to set the parent link pointer to null. Doing something like this doesn't work either:
```cpp
LinkSharedPtr parent_link = j->parent();
parent_link = nullptr;
```

Also, should this be an equality test or a serialization test? I believe @dellaert mentioned in some other PR or somewhere that either the joints should know about the links or the links should know about the joints and not both. We should probably find that.

Reply:
I was envisioning an equality test

or maybe something like
`*j->parent() = Link();`
?

Trying to make a false equality, preferably involving pointers.

Reply:
Okay this works. Seems like setting it shared ptr to nullptr doesn't quite work.